staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -176,6 +179,26 @@ func findPreexistingExpressionsInSchema(schema *apiextensions.JSONSchemaProps, e
 	})
 }
 
+// findMutatedVersions finds each version that is in the new CRD object and differs from that of the old CRD object.
+// It returns a set of the names of mutated versions.
+// This function does not check for duplicated versions, as further validations check their uniqueness.
+// This function does not care about the root version.
```

Comment:
why doesn't this function care about the root version? if there's only a single version, that's where it lives in the internal CRD type:

https://github.com/kubernetes/kubernetes/blob/ebf46ce1b45b8617d280b0aa12ebf6c18f25dea6/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion.go#L128-L169

Reply:
use the GetSchemaForVersion helper method to get the schema for the each version

Reply:
Thank you for the hint! main reason for not wanting to use `GetSchemaForVersion` is that it scans through versions, makes this function O(N^2) for N versions. However, because the number of versions are small, it is better to reuse existing helper function.